---
title: CA2000：超出范围前释放对象（代码分析）
description: 了解代码分析规则 CA2000：超出范围前释放对象
ms.date: 05/14/2019
ms.topic: reference
f1_keywords:
- CA2000
- Dispose objects before losing scope
- DisposeObjectsBeforeLosingScope
helpviewer_keywords:
- CA2000
- DisposeObjectsBeforeLosingScope
author: gewarren
ms.author: gewarren
dev_langs:
- CSharp
- VB
ms.openlocfilehash: 2e8f2d58ae8f8799fde51cc15522277cbc90988a
ms.sourcegitcommit: 05d0087dfca85aac9ca2960f86c5efd218bf833f
ms.translationtype: HT
ms.contentlocale: zh-CN
ms.lasthandoff: 03/30/2021
ms.locfileid: "99792422"
---
# <a name="ca2000-dispose-objects-before-losing-scope"></a>CA2000:丢失范围之前释放对象

| | 值 |
|-|-|
| **规则 ID** |CA2000|
| **类别** |[可靠性](reliability-warnings.md)|
| 修复是中断修复还是非中断修复 |非中断|

## <a name="cause"></a>原因

创建了 <xref:System.IDisposable> 类型的本地对象，但该对象不会被释放，除非对对象的所有引用都超出范围。

默认情况下，此规则会分析整个代码库，但这是[可配置](#configure-code-to-analyze)的。

## <a name="rule-description"></a>规则说明

如果在对某个可释放对象的所有引用超出范围之前未显式释放该对象，则当垃圾回收器运行该对象的终结器时，将在某个不确定的时间释放该对象。 由于可能发生异常事件，导致对象的终结器无法运行，因此应显式释放对象。

## <a name="special-cases"></a>特殊情况

即使未释放对象，也不会由于以下类型的本地对象触发规则 CA2000：

- <xref:System.IO.Stream?displayProperty=nameWithType>
- <xref:System.IO.StringReader?displayProperty=nameWithType>
- <xref:System.IO.TextReader?displayProperty=nameWithType>
- <xref:System.IO.TextWriter?displayProperty=nameWithType>
- <xref:System.Resources.IResourceReader?displayProperty=nameWithType>

将其中一个类型的对象传递给构造函数，然后将其分配给一个字段，表示释放所有权转移到新构造的类型。 也就是说，新构造的类型现在负责释放对象。 如果代码将其中一个类型的对象传递给构造函数，则即使在对对象的所有引用超出范围之前未释放该对象，也不会发生规则 CA2000 冲突。

## <a name="how-to-fix-violations"></a>如何解决冲突

要解决此规则的冲突，需在对对象的所有引用超出范围之前，在对象上调用 <xref:System.IDisposable.Dispose%2A>。

可使用 [`using` 语句](../../../csharp/language-reference/keywords/using-statement.md)（Visual Basic 中的 [`Using`](../../../visual-basic/language-reference/statements/using-statement.md)）来包装实现 <xref:System.IDisposable> 的对象。 以这种方式包装的对象将自动在 `using` 块的末尾释放。 但是，以下情况不应或不能使用 `using` 语句进行处理：

- 要返回可释放对象，该对象必须在 `using` 块外的 `try/finally` 块中构造。

- 请勿在 `using` 语句的构造函数中初始化可释放对象的成员。

- 如果构造函数仅由一个异常处理程序保护并嵌套在 [`using` 语句的获取部分](../../../csharp/language-reference/keywords/using-statement.md)，则外部构造函数中的失败会导致始终不会关闭嵌套构造函数所创建的对象。 在下面的示例中，<xref:System.IO.StreamReader> 构造函数中的失败会导致始终不会关闭 <xref:System.IO.FileStream> 对象。 在这种情况下，CA2000 标记为违反规则。

   ```csharp
   using (StreamReader sr = new StreamReader(new FileStream("C:\myfile.txt", FileMode.Create)))
   { ... }
   ```

- 动态对象应使用影子对象实现 <xref:System.IDisposable> 对象的释放模式。

## <a name="when-to-suppress-warnings"></a>何时禁止显示警告

在以下情况可禁止显示此规则发出的警告：

- 在对象上调用了一个调用 `Dispose` 的方法，例如 <xref:System.IO.Stream.Close%2A>
- 引发警告的方法返回包装对象的 <xref:System.IDisposable> 对象
- 分配方法没有释放所有权；也就是说，释放对象的责任将转移到在方法中创建并返回给调用方的另一个对象或包装器

## <a name="configure-code-to-analyze"></a>配置代码以进行分析

使用下面的选项来配置代码库的哪些部分要运行此规则。

- [排除特定符号](#exclude-specific-symbols)
- [排除特定类型及其派生类型](#exclude-specific-types-and-their-derived-types)

可以仅为此规则、为所有规则或为此类别（[可靠性](reliability-warnings.md)）中的所有规则配置这些选项。 有关详细信息，请参阅[代码质量规则配置选项](../code-quality-rule-options.md)。

[!INCLUDE[excluded-symbol-names](~/includes/code-analysis/excluded-symbol-names.md)]

[!INCLUDE[excluded-type-names-with-derived-types](~/includes/code-analysis/excluded-type-names-with-derived-types.md)]

## <a name="related-rules"></a>相关规则

- [CA2213:应释放可释放的字段](ca2213.md)

## <a name="example-1"></a>示例 1

如果要实现返回可释放对象的方法，可在没有 catch 块的情况下使用 try/finally 块来确保释放对象。 通过使用 try/finally 块，允许在故障点引发异常，并确保对象已释放。

在 OpenPort1 方法中，调用打开 ISerializable 对象 SerialPort 或对 SomeMethod 的调用可能会失败。 此实现会引发 CA2000 警告。

在 OpenPort2 方法中，将声明两个 SerialPort 对象并将其设置为 null：

- `tempPort`，用于测试方法操作是否成功。

- `port`，用于返回方法的值。

构造 `tempPort` 并在 `try` 块中打开，以及在相同的 `try` 块中执行任何其他所需的工作。 在 `try` 块的末尾，打开的端口分配给将返回的 `port` 对象，`tempPort` 对象设置为 `null`。

`finally` 块检查 `tempPort` 的值。 如果不为 null，则表示方法中的操作失败，`tempPort` 关闭以确保释放所有资源。 如果方法的操作成功，则返回的端口对象将包含打开的 SerialPort 对象，如果操作失败，则值为 null。

```csharp
public SerialPort OpenPort1(string portName)
{
   SerialPort port = new SerialPort(portName);
   port.Open();  //CA2000 fires because this might throw
   SomeMethod(); //Other method operations can fail
   return port;
}

public SerialPort OpenPort2(string portName)
{
   SerialPort tempPort = null;
   SerialPort port = null;
   try
   {
      tempPort = new SerialPort(portName);
      tempPort.Open();
      SomeMethod();
      //Add any other methods above this line
      port = tempPort;
      tempPort = null;

   }
   finally
   {
      if (tempPort != null)
      {
         tempPort.Close();
      }
   }
   return port;
}
```

```vb
Public Function OpenPort1(ByVal PortName As String) As SerialPort

   Dim port As New SerialPort(PortName)
   port.Open()    'CA2000 fires because this might throw
   SomeMethod()   'Other method operations can fail
   Return port

End Function

Public Function OpenPort2(ByVal PortName As String) As SerialPort

   Dim tempPort As SerialPort = Nothing
   Dim port As SerialPort = Nothing

   Try
      tempPort = New SerialPort(PortName)
      tempPort.Open()
      SomeMethod()
      'Add any other methods above this line
      port = tempPort
      tempPort = Nothing

   Finally
      If Not tempPort Is Nothing Then
         tempPort.Close()
      End If

   End Try

   Return port

End Function
```

## <a name="example-2"></a>示例 2

默认情况下，Visual Basic 编译器会检查溢出情况下的所有算术运算符。 因此，任何 Visual Basic 算术运算都可能引发 <xref:System.OverflowException>。 这可能会导致 CA2000 等规则出现意外冲突。 例如，以下 CreateReader1 函数会产生 CA2000 冲突，因为 Visual Basic 编译器正在为加法发出的溢出检查指令可能会引发导致 StreamReader 无法释放的异常。

要解决此问题，可在项目中禁用 Visual Basic 编译器发出溢出检查，也可修改代码，如以下 CreateReader2 函数所示。

若要禁用发出溢出检查，请在解决方案资源管理器中右键单击项目名称，然后单击“属性”。 依次单击“编译”和“高级编译选项”，然后检查“不做整数溢出检查”  。

:::code language="vb" source="snippets/vb/all-rules/ca2000-dispose-objects-before-losing-scope-vboverflow_1.vb":::

## <a name="see-also"></a>另请参阅

- <xref:System.IDisposable>
- [释放模式](../../../standard/garbage-collection/implementing-dispose.md)
